torchao/quantization/quantize_/workflows/int4/int4_tensor.py

```diff
@@ -294,13 +214,257 @@ def _(func, types, args, kwargs):
+        # the dim that int4 packing happens
+        if original_packing_dim in [0, 1]:
+            packing_dim = 0
+        else:
+            packing_dim = 1
+
+        block_size = self.block_size.copy()
+        block_size = [block_size[0] * block_size[1], block_size[2]]
+
+        qdata_shape = size.copy()
+        qdata_shape[packing_dim] //= 2
+        qdata = self.qdata.reshape(*qdata_shape)
+        sz_shape = []
+        for i in range(len(size)):
+            sz_shape.append(size[i] // block_size[i])
+        # scale and zero_point have reversed dimensions
+        sz_shape[0], sz_shape[1] = sz_shape[1], sz_shape[0]
+
+        scale = self.scale.reshape(*sz_shape)
+        zero_point = self.zero_point.reshape(*sz_shape)
+    elif len(original_shape) == 2 and len(size) == 3:
+        # only support extending the dim 0 to 2, `t.unflatten(0, (num_experts, -1))`
+        assert original_shape[-1] == size[-1], (
+            f"Only support reshaping when last dimension matches, requested: reshaping from {original_shape} to {size}"
+        )
+        if original_packing_dim == 0:
+            packing_dim = 1
+        else:
+            # original_packing_dim is 1
+            packing_dim = 2
+
+        block_size = self.block_size.copy()
+        block_size = [1, block_size[0], block_size[1]]
+
+        qdata_shape = size.copy()
+        qdata_shape[packing_dim] //= 2
+        qdata = self.qdata.reshape(*qdata_shape)
+
+        sz_shape = []
+        for i in range(len(size)):
+            sz_shape.append(size[i] // block_size[i])
+
+        # scale and zero_point have reversed dimensions
+        sz_shape[1], sz_shape[2] = sz_shape[2], sz_shape[1]
+
+        scale = self.scale.reshape(*sz_shape)
+        zero_point = self.zero_point.reshape(*sz_shape)
+    elif len(original_shape) == len(size):
+        assert all(x == y or y == -1 for x, y in zip(original_shape, size)), (
+            f"Only support viewing with match dimensions or -1, got: {original_shape}, {size}"
+        )
+        packing_dim = original_packing_dim
+        block_size = self.block_size
+    else:
+        assert len(original_shape) == 2 and len(size) == 3, (
+            f"Only support reshaping from 2D to 3D or from 3D to 2D or between sam ranges, requested: reshaping from {original_shape} to {size}"
+        )
+
+    shape = list(qdata.shape)
+    for i in range(len(shape)):
+        if i == packing_dim:
+            shape[i] *= 2
+
+    new = self.__class__(
+        qdata,
+        scale,
+        zero_point,
+        block_size,
+        shape,
+    )
+    return return_and_correct_aliasing(func, args, kwargs, new)
+
+
+@implements(aten.squeeze.dim)
+def _(func, types, args, kwargs):
+    self, dim = args
+
+    # Squeeze qdata
+    qdata = self.qdata.squeeze(dim=dim)
+
+    # For scale and zero_point, we need to squeeze based on the tensor layout
+    # Int4Tensor has scale and zero_point with shape (K/group_size, N) for 2D or (B, N, K/group_size) for 3D
```

Comment:
this is confusing, would you mind sharing some context on why this is true? why are the 2d and 3d cases different for the order of K/group_size and N stored on disk?